### DISCUSSION TITLE
casual dired on Windows and Git bash's `ls`

### DISCUSSION THREAD

**Discussion by dandrake** [kickingvegas/casual-dired#37] [Show and tell]
Here's my show-and-tell on getting casual-dired's sort-by stuff to work as expected on Windows.

First, my context:

* emacs 29 on Windows 10
* I use Git bash for most shell stuff; I'm using the usual [Git for Windows](https://git-scm.com/download/win) 2.43. 

When I first installed casua-dired, the sort-by menu didn't show up. That's because the variable `dired-use-ls-dired` was not set -- see `casual-dired-show-sort-by-tmenu-p`. I didn't know if the `ls` being used by dired/emacs supports it -- but I set that to `t` and things seemed to work.

My major workflows for sorting are by size or modification time, in either regular or reverse format. The questions I'm asking are "what file just got edited here?" or "what's the biggest file?" 

But, I found that the various date sorting wasn't working. I did a little digging, and because (1) I am kinda lazy, and (2) @kickingvegas is absurdly responsive and helpful, I filed an issue: #29. But then I thought I'd take a look at the code.

Looking at [the code](https://github.com/kickingvegas/casual-dired/blob/07eeb64dfe576e2926d49cf0a87801a579885785/lisp/casual-dired-sort-by.el#L74C26-L74C34,) the menu uses `casual-dired--sort-by-date-modified`. Okay, That calls into [this function](https://github.com/kickingvegas/casual-dired/blob/07eeb64dfe576e2926d49cf0a87801a579885785/lisp/casual-dired-sort-by.el#L168C9-L169C28) and there we see that, in previous versions of this package, when doing various date-related sorting, emacs and dired would execute something like `ls --dired --sort=time --time=modification`. But in git bash on my system, that gives:

```
$ ls --dired --sort=time --time=modification
ls: invalid argument 'modification' for '--time'
Valid arguments are:
  - 'atime', 'access', 'use'
  - 'ctime', 'status'
  - 'birth', 'creation'
Try 'ls --help' for more information.
```

Hrm. What kind of `ls` do I have on my system?

```
$ ls --version
ls (GNU coreutils) 8.32
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Richard M. Stallman and David MacKenzie.
```

The current version of this package, thanks to the fix to #29, now supports that older version of `ls`. But we are not done -- we need to make dired use that particular ls program because [dired uses an internal lisp implementation of ls by default on Windows](https://www.gnu.org/software/emacs/manual/html_node/efaq-w32/Dired-ls.html). So, following the instructions on that page, we do something like

```
(setq ls-lisp-use-insert-directory-program t)      ;; use external ls
(setq insert-directory-program "c:/cygwin/bin/ls") ;; ls program name
```

in our configuration. To use Git's `ls`, I have

```
(setq insert-directory-program "/Program Files/Git/usr/bin/ls")
```

Now, I can use this package's lovely menu interface to do sorting in dired without remembering all the weird dired keybindings. Nice!

**Comment by kickingvegas** [comment]
@dandrake I just released v1.1.0 with more GNU ls options support. Here's a screenshot.

https://github.com/kickingvegas/casual-dired/blob/9f65992e362c65a3023e4e2bc4853df9e92fa2d4/docs/images/casual-dired-sort-by-screenshot.png
